### PR TITLE
Update `#[coverage(..)]` attribute error messages to match the current implementation

### PR DESCRIPTION
**Pull request by Zalathar**
The allowed positions for `#[coverage(..)]` attributes were expanded by #126721, but the corresponding error messages were never updated to reflect the new behaviour.

Part of #134749.

### CODE REVIEW

File: compiler/rustc_passes/src/errors.rs [hunk 1]
```diff
@@ -71,13 +71,21 @@ pub(crate) struct InlineNotFnOrClosure {
...
+    /// "not a function, impl block, or module"
+    #[label(passes_not_fn_impl_mod)]
+    pub not_fn_impl_mod: Option<Span>,
+    /// "function has no body"
+    #[label(passes_no_body)]
+    pub no_body: Option<Span>,
+    /// "coverage attribute can be applied to a function (with body), impl block, or module"
+    #[help]
+    pub help: (),
 }
```
Comment (Zalathar) [inline compiler/rustc_passes/src/errors.rs:89]: Aside: `#[derive(Diagnostic)]` is one of the most egregiously unpleasant compiler-internal APIs I've encountered. It's *awful*.

File: compiler/rustc_passes/messages.ftl [hunk 2]
```diff
@@ -112,9 +112,11 @@ passes_coroutine_on_non_closure =
...
     .label = not a closure
 
-passes_coverage_not_fn_or_closure =
-    attribute should be applied to a function definition or closure
-    .label = not a function or closure
+passes_coverage_attribute_not_allowed =
+    coverage attribute not allowed here
+    .not_fn_impl_mod = not a function, impl block, or module
+    .no_body = function has no body
+    .help = coverage attribute can be applied to a function (with body), impl block, or module
```
Comment (jieyouxu) [inline compiler/rustc_passes/messages.ftl:119]: Question: does this specific wording account for this particular case:

```rs
// In situations where attributes can already be applied to expressions,
// the coverage attribute is allowed on closure expressions.
let _closure_tail_expr = {
    #[coverage(off)]
    || ()
};
```

File: compiler/rustc_error_codes/src/error_codes/E0788.md [hunk 3]
```diff
@@ -1,26 +1,23 @@
...
-For now, this attribute can only be applied to function, method, and closure
-definitions. In the future, it may be added to statements, blocks, and
-expressions, and for the time being, using this attribute in those places
-will just emit an `unused_attributes` lint instead of this error.
+The coverage attribute can be applied to:
+- Function and method declarations that have a body, including trait methods
+  that have a default implementation.
+- Closure expressions, in situations where attributes can be applied to
+  expressions.
+- `impl` blocks (inherent or trait), and modules.
```
Comment (jieyouxu) [inline compiler/rustc_error_codes/src/error_codes/E0788.md:9]: Question: I know this is very verbose for an error message, but cf. previous commit's error messages like "not a function, impl block, or module" or "coverage attribute can be applied to a function (with body), impl block, or module", the docs here doesn't quite correspond to those messages, right?

Anyway, not a huge deal.

File: compiler/rustc_error_codes/src/error_codes/E0788.md [hunk 4]
```diff
@@ -1,26 +1,23 @@
...
 
-`#[coverage(off)]` tells the compiler to not generate coverage instrumentation
-for a piece of code when the `-C instrument-coverage` flag is passed. Things
-like structs and consts are not coverable code, and thus cannot do anything
-with this attribute.
-
-If you wish to apply this attribute to all methods in an impl or module,
-manually annotate each method; it is not possible to annotate the entire impl
-with a `#[coverage]` attribute.
+The coverage attribute is a hint to the `-C instrument-coverage` flag to
```
Comment (jieyouxu) [inline compiler/rustc_error_codes/src/error_codes/E0788.md:20]: Nit: maybe also emphasis on "hint"? I.e. **hint** to the compiler.

**Review by jieyouxu — COMMENTED** [review]
Thanks, I have a few questions, but these error messages do indeed look much more accurate than it was previously.

File: compiler/rustc_passes/src/errors.rs (hunk 1 above)
Comment (jieyouxu) [reply compiler/rustc_passes/src/errors.rs:89]: cf. #132181

File: compiler/rustc_passes/messages.ftl (hunk 2 above)
Comment (Zalathar) [reply compiler/rustc_passes/messages.ftl:119]: My intention here is to make the error message less verbose by implicitly treating closure expressions as a kind of “function”, whereas the error code documentation has the luxury of listing closures separately.

Is this a good idea? I'm not sure; I don't have much experience with user-facing error messages.

**Comment by Zalathar** [comment]
Tweaked the wording of the error code docs [(diff)](https://github.com/rust-lang/rust/compare/94fa5b2851944f1c5920c55cd5bdf2740b43d1f5..e48fc62ce075985089f117e0472586be1af167b7).

I think the other stuff is good enough to merge as a net improvement over the current docs; if someone has strong editorial opinions they can raise them in a follow-up thread or PR.

@rustbot ready

File: compiler/rustc_passes/messages.ftl (hunk 2 above)
Comment (jieyouxu) [reply compiler/rustc_passes/messages.ftl:119]: I see what you mean. Should we perhaps say something like

> help: coverage attribute can be applied to a function with body, impl block, module, and a few other kinds of attachees; see error code docs for more details

But yeah, I agree that it's not very helpful to exhaustively list them either.

File: tests/ui/coverage-attr/allowed-positions.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
+//! Tests where the `#[coverage(..)]` attribute can and cannot be used.
+
+//@ reference: attributes.coverage.allowed-positions
```
Comment (jieyouxu) [inline tests/ui/coverage-attr/allowed-positions.rs:3]: cc @rust-lang/spec: should compiler reviewers ping you if a reference annotated test has expanded test coverage or modified test coverage?

File: compiler/rustc_passes/messages.ftl (hunk 2 above)
Comment (jieyouxu) [reply compiler/rustc_passes/messages.ftl:119]: I don't think this needs to be perfect (in this PR anyway).

**Review by jieyouxu — APPROVED** [review]
Thanks! The wording can always be adjusted in a follow-up, but the improvements are already nice.

File: tests/ui/coverage-attr/allowed-positions.rs (hunk 5 above)
Comment (jieyouxu) [reply tests/ui/coverage-attr/allowed-positions.rs:3]: Actually don't answer that (sorry), I'll add that as a question for the T-spec testing RFC.

**Pull request merged**